crates/wasmparser/src/validator/operators.rs

```diff
@@ -2156,122 +2167,107 @@ impl OperatorValidator {
+                        for ty in ft.inputs().rev() {
+                            self.pop_operand(Some(ty), resources)?;
                         }
-                        _ => {
-                            if !self.control.last().unwrap().unreachable {
-                                bail_op_err!(
-                                    "type mismatch: call_ref only works on index-type references"
-                                )
-                            }
+                        for ty in ft.outputs() {
+                            self.push_operand(ty, resources)?;
                         }
-                    }
+                    },
+                    HeapType::Bot => (),
+                    _ => bail_op_err!("type mismatch: instruction requires function reference type but stack has {}", ty_to_str(ValType::Ref(rt)))
                 }
                 self.check_return(resources)?;
             }
             Operator::RefAsNonNull => {
                 self.check_function_references_enabled()?;
-                if let Some(RefType { heap_type, .. }) = self.pop_ref(resources)? {
-                    self.check_heap_type(heap_type, resources)?;
-                    self.push_operand(
-                        ValType::Ref(RefType {
-                            nullable: false,
-                            heap_type,
-                        }),
-                        resources,
-                    )?;
-                }
+                let rt = self.pop_ref(resources)?;
+                self.push_operand(
+                    ValType::Ref(RefType {
+                        nullable: false,
+                        heap_type: rt.heap_type,
+                    }),
+                    resources,
+                )?;
             }
             Operator::BrOnNull { relative_depth } => {
                 self.check_function_references_enabled()?;
-                let (ty, kind) = self.jump(relative_depth)?;
-                let non_null = if let Some(RefType { heap_type, .. }) = self.pop_ref(resources)? {
-                    self.check_heap_type(heap_type, resources)?;
-                    ValType::Ref(RefType {
-                        nullable: false,
-                        heap_type,
-                    })
-                } else {
-                    // TODO: i'm confused. arbitrary but still tested as being
-                    // a ref?
-                    ValType::Ref(RefType {
-                        nullable: false,
-                        heap_type: HeapType::Func,
-                    })
-                };
-                // validates that t* matches block type by popping each t and
-                // pushing them again. TODO: This is not quite right with
-                // subtyping, and has to be changed everywhere
-                for ty in label_types(ty, resources, kind)?.rev() {
+                let rt = self.pop_ref(resources)?;
```

Comment:
Ah great, all fixed!  Much better, thanks!

I had some kind of point in that comment about how pop then push no longer preserves the types under subtyping, but the pseudocode in the spec uses it and it clearly passes the tests, so I must be wrong.   I might have to think about it to convince myself.